File: DataStructures/Hash/atcoder/RollingHash/A56/A56.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
)

const (
    MOD  = 1000000007
    BASE = 31
)

func main() {
    reader := bufio.NewReaderSize(os.Stdin, 1<<20)
    writer := bufio.NewWriterSize(os.Stdout, 1<<20)
    defer writer.Flush()

    // 入力読み取り
    var N, Q int
    fmt.Fscan(reader, &N, &Q)

    var S string
    fmt.Fscan(reader, &S)

    // 前計算
    hash := make([]int64, N+1)   // hash[i] = S[0..i-1] のハッシュ
    power := make([]int64, N+1)  // BASE^i の累乗

    power[0] = 1
    for i := 0; i < N; i++ {
        code := int64(S[i] - 'a' + 1)
        hash[i+1] = (hash[i]*BASE + code) % MOD
        power[i+1] = (power[i] * BASE) % MOD
    }

    // クエリ処理
    scanner := bufio.NewScanner(reader)
    scanner.Split(bufio.ScanWords)

    nextInt := func() int {
        scanner.Scan()
        val, _ := strconv.Atoi(scanner.Text())
        return val
    }

    for i := 0; i < Q; i++ {
        a := nextInt()
        b := nextInt()
        c := nextInt()
        d := nextInt()

        hash1 := getHash(hash, power, a, b)
        hash2 := getHash(hash, power, c, d)

        if hash1 == hash2 {
            fmt.Fprintln(writer, "Yes")
        } else {
            fmt.Fprintln(writer, "No")
        }
    }
}

func getHash(hash, power []int64, l, r int) int64 {
    res := hash[r] - (hash[l-1]*power[r-l+1])%MOD
    if res < 0 {
        res += MOD
    }
    return res
}

File: DataStructures/Hash/atcoder/RollingHash/A56/A56.php
<?php
fscanf(STDIN, "%d %d", $N, $Q);
$S = trim(fgets(STDIN));

define('MOD', 1000000007);
define('BASE', 31);

// プレフィックスハッシュと累乗を初期化
$hash = array_fill(0, $N + 1, 0);
$power = array_fill(0, $N + 1, 1);

// ハッシュと累乗の前計算
for ($i = 0; $i < $N; $i++) {
    $code = ord($S[$i]) - ord('a') + 1;
    $hash[$i + 1] = ($hash[$i] * BASE + $code) % MOD;
    $power[$i + 1] = ($power[$i] * BASE) % MOD;
}

// ハッシュ取得関数（1-indexed）
function getHash($l, $r, $hash, $power)
{
    $val = ($hash[$r] - ($hash[$l - 1] * $power[$r - $l + 1]) % MOD + MOD) % MOD;
    return $val;
}

// クエリ処理
for ($i = 0; $i < $Q; $i++) {
    fscanf(STDIN, "%d %d %d %d", $a, $b, $c, $d);
    $h1 = getHash($a, $b, $hash, $power);
    $h2 = getHash($c, $d, $hash, $power);
    echo ($h1 === $h2 ? "Yes" : "No") . "\n";
}